test/runner/job_test.exs

```diff
@@ -45,4 +19,253 @@ defmodule ElixirBench.Runner.JobTest do
     assert job.context.worker_num_cores in 1..64
     assert job.context.worker_os in [:Linux, :Windows, :macOS]
   end
+
+  describe "start_job/6" do
+    test "run job and return results given execute function" do
+      task_fun = fn job ->
+        {:ok, %{job | log: "success", status: 0}}
+      end
+
+      {:ok, job} = Job.start_job(job_fixture(), task_fun: task_fun)
+
+      assert %{log: "success", status: 0} = job
+    end
+
+    test "not start job if there is other job running" do
+      task_fun = fn _job ->
+        :timer.sleep(2_000)
+        {:ok, %{}}
+      end
+
+      %{pid: pid} =
+        Task.Supervisor.async_nolink(ElixirBench.Runner.JobsSupervisor, fn ->
+          :timer.sleep(2_000)
+        end)
+
+      assert_raise MatchError, fn ->
+        Job.start_job(job_fixture(), task_fun: task_fun)
+      end
+
+      Process.exit(pid, :kill)
+    end
+
+    test "handle job timeout" do
+      task_fun = fn _job ->
+        :timer.sleep(2_000)
+        {:ok, %{}}
+      end
+
+      job = Job.start_job(job_fixture(), task_fun: task_fun, timeout: 100)
+
+      assert %{log: "Job execution timed out", status: 127} = job
+    end
+  end
+
+  describe "ensure_no_other_jobs!/0" do
+    test "raise error if active child process already exist under JobsSupervisor" do
```

Comment:
this might be controversial, as I don't think anyone else calls it I think it should be tested as part of `start_job` and this is just an implementation detail and keep the method private.

That gives us a smaller API surface and less possible rework due to API changes as we could completely remove the method as long as we keep the functionality

Reply:
I agree with you. But the job run is very complex and making methods public with `@doc false` sounds like a good solution in a way we can test the feature and corner cases without complex and time consuming tests, calling just `Job.start_job`. I got inspired in some discussions and posts like this one '[testing private functions](https://pragdave.me/blog/2017/03/31/tesing-private-functions.html)'  from PragDave, that uses a  library to make methods public under test environment. Another approach which I've found more interesting given its simplicity, is to make it public and add `@doc false` tags. Anyone can call it from the outside scope of the module but for me its a good signature and it also does not appear as an API/public method in the console, so I very liked it.